### ISSUE TITLE
time doesn't appear correctly when editing a "once-per-day" playlist

### ISSUE THREAD

**Issue by cpeter1207** [AzuraCast/AzuraCast#862]
**Installation method**
traditional

**Host OS (for Docker installations)**
Ubuntu Xenial

**Describe the bug**
1. Set your timezone to CDT
2. Create a "once per day" playlist scheduled for Saturday at 7:00 PM CDT.
3. Click "save". the playlist appears correctly in the schedule view.
4. Edit the playlist. The time will be displayed as 12:00 PM CDT (5:00 AM UTC). This is not correct. It should be displayed as 7:00 PM CDT (12:00 AM CDT).

I also noticed two other issues while investigating this:

1. While the playlist was scheduled for Saturday (the Saturday checkbox was checked), it played on Friday at 7:06 PM CDT (Saturday at 12:00 AM UTC).

2. When looking at liquidsoap.liq, the entry in the switch statement associated with this playlist did not have any day-of-week qualifiers. In other words, the time component was simply 19h. I think it should be 6w and 19h if I've only checked the "saturday" checkbox.

I'm a professional software developer, by day, and I'm interested in contributing to this project in my spare time. I haven't learned enough about your code to understand where to start investigating these issues, but I'm willing to learn and to contribute patches if you can give me some education abouthow this code is organized.

I primarily work in java, so I'll need to learn PHP also, but I think I should be able to pick it up quickly if I have specific tasks to work on. I do have some experience with manually scripting liquidsoap, though so I'm currently working around this by manually adding the schedule for this playlist using liquidsoap.

I'd be willing to investigate this issue further as my first task if you're interested in having me contribute. I suspect that it's a pretty straight-forward timezone conversion issue.

**Comment by SlvrEagle23** [comment]
@cpeter1207 Thank you for the detailed report.

The first part of your report, that selecting the 7:00pm CDT / 12:00am UTC time doesn't persist to the next time you edit the form, was indeed a valid issue and has been fixed as of the most recent code update. I also was able to fix an unrelated issue with failed validation causing some fields to not update.

The second part, about the time actually properly resolving into Liquidsoap configuration, is substantially more complicated.

For one, the Liquidsoap configuration notation is only intended as a fallback in the event AzuraCast's own AutoDJ fails to produce a song to play next. The AutoDJ's code for checking whether a song should play next is [here](https://github.com/AzuraCast/AzuraCast/blob/master/src/Radio/AutoDJ.php#L183-L207) and [here](https://github.com/AzuraCast/AzuraCast/blob/master/src/Entity/StationPlaylist.php#L496-L510), and that's usually the authoritative source for what Liquidsoap should be playing.

Secondly, this particular issue is a beast because there's an existential question at play: when you say "Saturday" do you mean the CDT Saturday, which would play at what would be 12am Sunday in UTC, or do you mean UTC Saturday, which would play at what would be 7pm Friday in CDT? Currently we use the latter interpretation, though it may be more intuitive to use the former.

Time zone management is the least fun thing ever. I'll leave this issue open for discussion and possible future resolution of the issue.

**Comment by SlvrEagle23** [comment]
@cpeter1207 Now I remember why I went with the second, perhaps less intuitive interpretation of meaning the UTC day, not the local day: because when determining the time something should play, I don't know what the user's local time zone was.

The real solution here is to adopt something akin to AirTime's way of dealing with this, which is to allow per-playlist time zone setting, so you can clearly indicate on each playlist what zone the times are relative to. This time zone would default to your local one, but could be changed at any time.

I'm planning on doing something of a heavy overhaul of this system in the near future, pending the resolution of the issues mentioned in #860. I would expect a resolution to this to accompany those changes.

**Comment by cpeter1207** [comment]
@SlvrEagle23 

Thanks for the quick response and the commit addressing the first part of my issue. I generally found airtime to be difficult to deal with in most respects, but it might be most intuitive to let me specify the timezone in which the day is specified like I can with the time.

I could invision some logic in the form that converts the local days to UTC automagically based on the timezone I specified in the time field. The logic might look something like:

if localtime < gmttime then gmtday = localday-1
else if localtime == gmttime then gmtday = localday
else gmtday = localday+1

It appears that, whether you intended or not, the code that writes liquidsoap.liq does convert the times into the correct timezone for the machine (which, in my case is CDT), so something like the reverse of the above would allow you to write the days to liquidsoap.liq correctly to address the last part of my issue.

I understand, though, that the liquidsoap code is really just a fallback, so that was more of an observation than an actual issue.

I think the thing I'm most concerned about, at the moment, is that the day selections on the form follow the timezone I specified the time in so it's clear to me that I've scheduled the playlist correctly.

The second most thing I'm concerned about is trying to get a better understanding of this code since, as I said, I'm interested in contributing. Given the other post you referenced, contributing seems like it could be helpful to other users, if not directly helpful to your situation.

Having suffered with depression myself, I know that it's sometimes helpful, at least in the short-term, to kick around ideas about how to solve interesting problems, so if I can be helpful to you in that way, I'm glad to do so.
